torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp

```diff
@@ -24,36 +24,38 @@ using namespace ::torch::jit;
 // conversion module into a group. We should re-implement this.
 std::vector<Node*> FakeCluster(const std::shared_ptr<Graph>& graph) {
   std::vector<Node*> nodes;
-  for (auto node : graph->nodes()) {
-    if (torch::blade::IsMlirMhloSupported(*node) &&
-        node->kind() != prim::Constant) {
-      if (node->kind() == aten::addmm) continue;
-      nodes.push_back(node);
-    }
-  }
+  std::copy_if(graph->nodes().begin(), graph->nodes().end(),
+               std::back_inserter(nodes), [](torch::jit::Node* node) {
+                 return torch::blade::IsMlirMhloSupported(*node) &&
+                        node->kind() != prim::Constant;
+               });
   return nodes;
 }
 
 // Give:
 //  with prim::FusionGroup(
-//      %1: Scalar,
-//      %2: Scalar):
-//    %3 Tensor = aten::add(%1, %2)
-//  return %3
+//      %p0: Tensor,
+//      %p1: Tensor,
+//      %p2: Scalar):
+//    %1 Tensor = aten::add(%p0, %p1, %p2)
+//  return %1
 //
 // Execute: CastToTensorInputs(sub_graph)
 //
 // After:
 //  with prim::FusionGroup(
-//      %1.1: Tensor,
-//      %2.1: Tensor):
-//    %4 = aten::item(%1.1, 1)
-//    %5 = aten::item(%2.1, 1)
-//    %3 Tensor = aten::add(%4, %5)
-//    return %3
+//      %p0.1: Tensor,
+//      %p1.1: Tensor,
+//      %p2.1: Tensor):
+//    %1 int = aten::item(%p2.1)
+//    %2 Tensor = aten::add(%p0.1, %p1.1, %1)
+//    return %2
 void CastBoundaryScalarToTensor(Graph* disc_graph, size_t i) {
   auto new_input = disc_graph->insertInput(
       i, c10::string(disc_graph->inputs()[i]->debugName() + ".1"));
+  // TODO(disc): derivate input scalar type
+  new_input->setType(
+      TensorType::create(at::ScalarType::Int, c10::nullopt, 0, false));
```

Comment:
Is this pass currently only used in the experimental training support? (otherwise blindly setting dtype to int would affect correctness of other cases)

Reply:
thanks for your reminder, It's not only the experimentation, and hard code to int would cause unexpected behavior and be hard to debug, I will update this PR later. :)

Reply:
done, add `DerivateScalarType` function.

Reply:
done, added stricter inspection on input dtype and scalar type derivation using `c10::scalarTypeFromJitType`.